New contract method execution option: generating the transaction dictionary

monokh opened the pull request:
### What was wrong?

There is no convenient way to prepare a contract transaction for offline signing.

### How was it fixed?

Add a method to the contract api `prepareTransaction` to return a dictionary for easy transaction signing.

### Tasks

- [x] choose naming
- [x] implement new API
- [x] document in docs/
- [x] reference this method in signTransaction docs docs/web3.eth.account.rst

#### Cute Animal Picture
![Mandatory fluff ball](https://i.pinimg.com/originals/9b/49/7e/9b497ece12df1a9ff742485f67c4276c.jpg)

Fixes #430

monokh commented:
@carver The thing which is unclear for me is how we should infer some of the parameters.
Do we expect to be connected to a provider when `prepareTransaction` is called?

- **Gas:** We can use the `estimateGas` function to infer this
- **gasPrice:** Should we connect to the provider and get a default gas price? What if the provider is not available? Should we set a default like 20gwei? 
- **Nonce:** Same as gasPrice, i guess we would need connection to a provider and an account to provide the real nonce... 
- **chainId:** Can use `net_version ` rpc call although has to be connected to a network.

carver commented:
> Do we expect to be connected to a provider when `prepareTransaction` is called?

@monokh yes. We're doing an online preparation for an offline signing.

carver commented:
@monokh Feel free to @ me and/or remove "WIP" from the PR title when you're ready for a review

monokh commented:
@carver sure. Still want to add some more test cases and docs but do you think the general approach looks ok?

monokh commented:
Removed WIP, extended the tests and docs @carver hopefully its roughly what you are looking for 😃

pipermerriam commented on docs/contracts.rst:
```diff
@@ -172,6 +172,35 @@ Each Contract Factory exposes the following methods.
+        ``myMethod(*args, **kwargs)``.
+
+    Returns a transaction dictionary. This dictionary may be used for offline 
+    transaction signing using :meth:`~web3.eth.account.Account.signTransaction`.
```
Maybe the first example should be the simplest one which would be sending the transaction using `web3.eth.sendTransaction`.  Then you could mention the *fancier* version using `web3.eth.account.Account.signTransaction`

pipermerriam commented on docs/contracts.rst:
```diff
@@ -172,6 +172,35 @@ Each Contract Factory exposes the following methods.
+    function call, and function arguments being passed into the second portion.
+
+    .. note::
+        If ``web3.eth.defaultAccount`` is not set, this method will need another way
```
I think it it's ok to return a transaction dict without a `nonce`.  In fact, I'm inclined to have `nonce` inclusion be opt-in, here's my reasoning.

* If I'm preparing a *generic* transaction that I want to send to many different contracts, I don't want `nonce` included.
* If I'm preparing a transaction to be sent once, not including nonce won't hurt because the `web3.eth.sendTransaction` API (and underlying `eth_sendTransaction` JSON-RPC) automatically add the proper nonce.

I think that fully excluding `nonce` from the return dict unless it's included in the `transaction` parameter would be just as easy to work with and would make using this API with different use cases easier.

carver replied:
> i'm a little concerned how this would look for ...  prepare a single transaction, sign it and send it raw

This is a good instinct. We want to make sure the simple things stay easy. On the other hand, we risk confusing users who don't think about nonces explicitly.

Problems
---------

If someone prepares two transactions for an account, the second will clobber the first transaction, or they will end up with a confusing error about gas price not being high enough. For example:

```
transaction_dict1 = myContract.functions.myMethod().asdict()
transaction_dict2 = myContract.functions.myMethod().asdict()

signed = w3.eth.account.signTransaction(transaction_dict1, private_key)
w3.eth.sendRawTransaction(signed.rawTransaction)

signed = w3.eth.account.signTransaction(transaction_dict2, private_key)
w3.eth.sendRawTransaction(signed.rawTransaction)
# ^ error or overwrite
```

Note that with the current implementation, this scenario even breaks for the client-signed approach, like:

```
transaction_dict1 = myContract.functions.myMethod().asdict()
transaction_dict2 = myContract.functions.myMethod().asdict()

w3.eth.sendTransaction(transaction_dict1)
w3.eth.sendTransaction(transaction_dict2)
# ^ error or overwrite
```

A Weak Solution
---------

One solution here is to add another public "pre-signing" method to always be called immediately before offline signing. Something like:

```
transaction_dict1 = myContract.functions.myMethod().asdict()
transaction_dict2 = myContract.functions.myMethod().asdict()

# toSignableTransaction must be called immediately before signing and sending,
# ... and called after the most recently sent transaction

signable_transaction = w3.eth.toSignableTransaction(transaction_dict1)
signed = w3.eth.account.signTransaction(signable_transaction, private_key)
w3.eth.sendRawTransaction(signed.rawTransaction)

signable_transaction = w3.eth.toSignableTransaction(transaction_dict2)
signed = w3.eth.account.signTransaction(signable_transaction, private_key)
w3.eth.sendRawTransaction(signed.rawTransaction)
```

It adds an extra step, but might save us some headaches with people who prepare transactions and don't sign/send them immediately.

`toSignableTransaction` still isn't trivial in this case, because `getTransactionCount` will not recognize the pending transactions afaik, so we'd have to be even more clever about tracking the pending transactions.

A Better Solution
----------------

Explicitly calling `getTransactionCount` is actually the best option I can see. That's because tracking nonces is so hard that we should not even try to do it automatically, we should ask the user to figure it out explicitly, like:

```
transaction_dict1 = myContract.functions.myMethod().asdict()
transaction_dict2 = myContract.functions.myMethod().asdict()

transaction_dict1['nonce'] = w3.eth.getTransactionCount(transaction_dict1['from'])
transaction_dict2['nonce'] = transaction_dict1['nonce'] + 1

signed = w3.eth.account.signTransaction(transaction_dict1, private_key)
w3.eth.sendRawTransaction(signed.rawTransaction)

signed = w3.eth.account.signTransaction(transaction_dict2, private_key)
w3.eth.sendRawTransaction(signed.rawTransaction)
```

Then this will work transparently again, without deleting the auto-generated nonce values:

```
transaction_dict1 = myContract.functions.myMethod().asdict()
transaction_dict2 = myContract.functions.myMethod().asdict()

w3.eth.sendTransaction(transaction_dict1)
w3.eth.sendTransaction(transaction_dict2)
```

A Stretch to get the Best of Both Worlds
-------------
Maybe we can still get a clean single solution with a special keyword, like:

```
transaction_dict = myContract.functions.myMethod().asdict({'nonce': 'current'})

signed = w3.eth.account.signTransaction(transaction_dict, private_key)
w3.eth.sendRawTransaction(signed.rawTransaction)
```

Hopefully `current` is enough of a visual signal to people that two consecutive calls will be broken. (I'm not convinced about this part, I'm going to let this marinate and come back to it)

monokh replied:
Ah yes, i hadn't really considered tracking nonces... I think you are definitely right in that we should not even attempt it. It will only introduce more magic that is hard to figure out... 

In your "Better Solution" nonce can also be specified inline as a shortcut:
```
transaction_dict1 = myContract.functions.prepareTransaction({
'nonce':  w3.eth.getTransactionCount(myAccount)
}).myMethod()
```
The good thing about this is that it's pretty clearly presentable in an example with an extra mention of how nonce management is left upto the user. 
I think we are all in agreement then that for now nonce should not be handled by the method. I will make that change.

pipermerriam replied:
How about instead of special keyword magic we just allow an extra kwarg to that API:

```python
transaction_dict = myContract.functions.myMethod().asdict({'from': web3.eth.coinbase}, add_nonce=True)
```

pipermerriam replied:
(duplicating comment here)

How about instead of special keyword magic we just allow an extra kwarg to that API:

```python
transaction_dict = myContract.functions.myMethod().asdict(
    {'from': web3.eth.coinbase},
    add_nonce=True,
)
```

monokh replied:
Happy to add that @pipermerriam, was the previous behaviour of how nonce was added suitable (only with this extra flag ofcourse). sorry for wiping that out pre-emptively :slightly_frowning_face: As reminder, it was
1. Nonce of provider default account
2. Nonce of account specified in call 'from' attribute
Also we need to then explain in docs that there will be potential nonce conflicts with this feature

carver replied:
@monokh Rather than go back and forth on how to specify auto-nonce, let's just leave it out altogether. We can come back to it later in another issue, if it becomes important. You're example explicitly using `getTransactionCount` looks good.

carver replied:
I'm pretty sure his comment was "if auto-nonce stays in, maybe this way would look better", rather than "I'm advocating for auto-nonce to go back in".

carver replied:
Oh and that means the docs should change too. `from` or `nonce` no longer need to be specified here.

monokh replied:
Ok, let me know if you would like to revisit this one :) happy to help with this in another PR

carver commented on tests/core/contracts/test_contract_prepareTransaction.py:
```diff
@@ -0,0 +1,115 @@
+    return _math_contract
+
+
+def test_prepare_transaction_with_contract_no_arguments(web3, math_contract):
```
You could do a duplicate of the test at line 18, but add the MathContract fixture. Then you can call:

```py
MatchContract.prepareTransaction({..., 'to': math_contract.address})
```

carver commented on tests/core/contracts/test_contract_prepareTransaction.py:
```diff
@@ -0,0 +1,115 @@
+        'Explicit Gas',
+        'Explicit Gas Price',
+        # 'Explicit Nonce',  # eth-testrpc sendTransaction breaks when nonce is set
+        #  https://github.com/pipermerriam/eth-testrpc/issues/98
```
This is sort of a future-proofing thing. At some point, we could switch to an EthereumTester and/or py-evm backend, and we can turn this test back on easily.

carver commented on web3/contract.py:
```diff
@@ -820,3 +871,27 @@ def estimate_gas_for_function(contract=None,
+
+    prepared_transaction = fill_transaction_defaults(contract.web3, prepared_transaction)
+    # from is not a valid transaction part and used only for retreiving nonce
+    prepared_transaction.pop('from')
```
If it does throw an exception (seems plausible), it might be reasonable to actually alter the `eth.account` code to only throw an exception if `from` is different from the key that's signing the transaction. If you confirm the exception, but aren't comfortable making that change, let me know and I'm happy to add it in.

monokh replied:
Ok i definitely made that up 😄  Anyway with the removal of nonce defaulting, `from` won't have to be passed into this method anyway.

monokh commented:
Ok :slightly_smiling_face: removed nonce defaulting and updated tests and docs as requested. How is it looking now @carver @pipermerriam ?
Travis seems to have given up though :(

carver commented:
LGTM I'll probably wait till Monday morning to merge this in case Piper wants to chime in any more.

pipermerriam commented on tests/core/contracts/test_contract_buildTransaction.py:
```diff
@@ -0,0 +1,124 @@
+    with pytest.raises(ValueError):
+        math_contract.buildTransaction({
+            'data': '0x000'
+        }).increment()
```
TODO: I think we should remove this call to `increment()` since I *think* the error occurs within `buildTransaction`?

monokh replied:
Yes it does :slightly_smiling_face: good catch. I can submit a follow up PR? Let me know if there's anything else to add

carver replied:
A new PR is the way to go :+1:
